torch/distributed/fsdp/_init_utils.py

```diff
@@ -22,6 +55,219 @@
 PARAM_BROADCAST_BUCKET_SIZE = int(250 * 1024 * 1024)
 FSDP_SYNCED = "_fsdp_synced"
 
+# TODO (awgu): Refactor this later
+SHARDING_STRATEGY_MAP = {
+    ShardingStrategy.NO_SHARD: HandleShardingStrategy.NO_SHARD,
+    ShardingStrategy.FULL_SHARD: HandleShardingStrategy.FULL_SHARD,
+    ShardingStrategy.SHARD_GRAD_OP: HandleShardingStrategy.SHARD_GRAD_OP,
+}
+
+
+# NOTE: Since non-self attributes cannot be type annotated, several attributes
+# on `state` are defined first as local variables before being assigned.
+
+
+@no_type_check
+def _init_process_group_state(
+    state: _State,
+    process_group: Optional[dist.ProcessGroup],
+) -> _State:
+    state.process_group = process_group or _get_default_group()
+    state.rank = state.process_group.rank()
+    state.world_size = state.process_group.size()
+    return state
+
+
+@no_type_check
+def _init_ignored_module_states(
+    state: _State,
+    module: nn.Module,
+    ignored_modules: Optional[Iterable[torch.nn.Module]],
+) -> _State:
+    state._ignored_modules = _get_ignored_modules(module, ignored_modules)
+    state._ignored_params, state._ignored_param_names = _get_ignored_params(
+        module,
+        state._ignored_modules,
+    )
+    # TODO: FSDP's contract for buffers is not well-defined. They are
+    # implicitly ignored for most functionality since they are not sharded;
+    # however, FSDP still imposes some semantics on buffers (e.g. buffer mixed
```

Comment:
buffer is a bit hard, as we cannot anticipate how users gonna use it. It could be masks, stats, etc. We might need users to provide some hook/func to handle it.

Reply:
Let's create an issue to track, but I think this is not P0 for now.

Reply:
Also, DDP has the same issue. The new contract can cover all composable APIs.